src/main/java/uk/ac/cam/cl/dtg/segue/comm/MailGunEmailManager.java

```diff
@@ -0,0 +1,180 @@
+                                           @Nullable final Map<Long, Map<String, Object>> userVariablesOrNull)
+            throws FeignException {
+
+        // Lazily construct the MailGun messages API
+        this.createMessagesApiIfNeeded();
+
+        // Ensure all relevant email variable maps are not null
+        Map<Long, Map<String, Object>> userVariables = null != userVariablesOrNull ? userVariablesOrNull : Map.of();
+        Map<String, Object> templateVariables = new HashMap<>(null != templateVariablesOrNull ? templateVariablesOrNull : Map.of());
+        // Combine local and global template variables (non-recipient ones)
+        if (null != this.globalStringTokens) {
+            templateVariables.putAll(this.globalStringTokens);
+        }
+
+        // This expression:
+        //  - filters users by their email preferences and email verification status
+        //  - for each user U, extracts the name and email of U, and adds them to the (possibly empty) map of
+        //    `userVariables.get(U.getId())`
+        //  - creates a new map with emails as keys and the updated `userVariables` maps as values
+        Map<String, Object> recipientVariables = userDTOs.stream()
+                .filter(user -> {
+                    // FIXME - should system emails ignore this setting to avoid abuse?
+                    if (user.getEmailVerificationStatus() == EmailVerificationStatus.DELIVERY_FAILED) {
+                        return false;
+                    }
+                    try {
+                        UserPreference preference = userPreferenceManager.getUserPreference(
+                                SegueUserPreferences.EMAIL_PREFERENCE.name(), emailType.name(), user.getId());
+                        // If no preference is present, do not send email to this user.
+                        return preference != null && preference.getPreferenceValue();
+                    } catch (SegueDatabaseException e) {
+                        // Assume the worst and do not send the email to this user in the case of a DB issue
+                    }
+                    return false;
+                }).collect(
+                    Collectors.toMap(
+                        RegisteredUserDTO::getEmail,
+                        u -> Stream.of(
+                                ImmutableMap.of(
+                                "email", u.getEmail(),
+                                "givenName", u.getGivenName()
+                                ),
+                                userVariables.getOrDefault(u.getId(), ImmutableMap.of())
+                        ).flatMap(m -> m.entrySet().stream())
+                        .collect(Collectors.toMap(Entry::getKey, Entry::getValue, (e1, e2) -> e1))
+                    )
+                );
+
+        String fromAddress = emailContentTemplate.getOverrideFromAddress();
+        if (null == fromAddress || fromAddress.isEmpty()) {
```

Comment:
Elsewhere we have used `com.google.common.base.Strings` and `Strings.isNullOrEmpty(...)`, which is slightly simpler but may not be worth it.

There's also `defaultIfBlank` [from the Apache Commons library](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#defaultIfBlank-T-T-) we have, to compress the whole lot, but then the lines would be really long.

May not be worth it!

Reply:
The more declarative the better!